packages/server/src/core/types.ts

```diff
@@ -24,22 +35,38 @@ export const procedureTypes = ['query', 'mutation', 'subscription'] as const;
 export type ProcedureType = typeof procedureTypes[number];
 
 export type inferHandlerInput<TProcedure extends Procedure<any>> =
-  ProcedureArgs<inferProcedureParams<TProcedure>>;
-
-export type inferProcedureInput<TProcedure extends Procedure<any>> =
-  inferHandlerInput<TProcedure>[0];
-
-// /**
-//  * @internal
-//  */
-// type inferProcedureFn<TProcedure extends Procedure<any>> =
-//   TProcedure extends QueryProcedure<any>
-//     ? TProcedure['query']
-//     : TProcedure extends SubscriptionProcedure<any>
-//     ? TProcedure['subscription']
-//     : TProcedure extends MutationProcedure<any>
-//     ? TProcedure['mutate']
-//     : never;
+  ProcedureArgs<TProcedure['_def']>;
+
+type RouterRecordValueShape<
+  TRouterOrProcedure extends AnyRouter | AnyProcedure,
+> = TRouterOrProcedure extends TRPCRouterItem<ItemKind.Router>
+  ? RouterShape<TRouterOrProcedure['_def']['record']>
+  : TRouterOrProcedure extends TRPCRouterItem<ItemKind.Procedure>
+  ? {
+      kind: TRouterOrProcedure['_procedureKind'];
+      input: ProcedureArgs<TRouterOrProcedure['_def']>;
+      output: TRouterOrProcedure['_def']['_output_out'];
```

Comment:
Notice the addition of the `kind`, which is required if you want to allow people to create types based on the procedure type.